phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/data/WalletPayment.kt

```diff
@@ -0,0 +1,136 @@
+ *
+ * It is common to reference these rows in other database tables via:
+ * - type: Long (representing either incoming or outgoing)
+ * - id: String (representing the appropriate id for either table)
+ *
+ * The WalletPaymentId class assists in this conversion.
+ */
+sealed class WalletPaymentId {
+
+    data class IncomingPaymentId(val paymentHash: ByteVector32): WalletPaymentId() {
+        companion object {
+            fun fromString(id: String) =
+                IncomingPaymentId(paymentHash = ByteVector32(id))
+
+            fun fromByteArray(id: ByteArray) =
+                IncomingPaymentId(paymentHash = ByteVector32(id))
+        }
+    }
+
+    data class OutgoingPaymentId(val id: UUID): WalletPaymentId() {
+        companion object {
+            fun fromString(id: String) =
+                OutgoingPaymentId(id = UUID.fromString(id))
+        }
+    }
+
+    enum class DbType(val value: Long) {
+        INCOMING(1),
+        OUTGOING(2)
+    }
+
+    /* Maps to database column:
+     * - type: INTEGER NOT NULL (mapped to Long via SqlDelight)
+     */
+    val dbType: DbType get() = when (this) {
+        is IncomingPaymentId -> DbType.INCOMING
+        is OutgoingPaymentId -> DbType.OUTGOING
+    }
+
+    /* Maps to database column:
+     * - id: TEXT NOT NULL (mapped to String via SqlDelight)
+     */
+    val dbId get() = when (this) {
+        is IncomingPaymentId -> this.paymentHash.toHex()
+        is OutgoingPaymentId -> this.id.toString()
+    }
+
+    /* Use `identifier` if the code requires a single (hashable) identifier for the row.
+     * For example, for use within a hashmap or Cache.
+     */
+    val identifier: String get() = when(this) {
+        is OutgoingPaymentId -> "outgoing|${this.id}"
+        is IncomingPaymentId -> "incoming|${this.paymentHash.toHex()}"
+    }
+
+    companion object {
+
+        fun create(type: Long, id: String): WalletPaymentId? {
+            return when(type) {
+                DbType.INCOMING.value -> {
+                    IncomingPaymentId.fromString(id)
+                }
+                DbType.OUTGOING.value -> {
+                    OutgoingPaymentId.fromString(id)
+                }
+                else -> null
+            }
+        }
+    }
+}
+
+fun WalletPayment.walletPaymentId(): WalletPaymentId = when (this) {
+    is IncomingPayment -> WalletPaymentId.IncomingPaymentId(paymentHash = this.paymentHash)
+    is OutgoingPayment -> WalletPaymentId.OutgoingPaymentId(id = this.id)
+}
+
+/**
+ * Represents a payment & its associated metadata.
+ */
+data class WalletPaymentInfo(
+    val payment: WalletPayment,
+    val metadata: WalletPaymentMetadata
+) {
+    fun id() = payment.walletPaymentId()
+}
+
+/**
+ * Represents information from the `payments_metadata` table.
+ */
+data class WalletPaymentMetadata(
+    val lnurl: LNUrl? = null,
+    val lnurlDescription: String? = null,
+    val lnurlSuccessAction: LNUrl.PayInvoice.SuccessAction? = null,
```

Comment:
How about wrapping the lnurl data in a nullable object, instead of having many nullable attributes?

```kotlin
data class WalletPaymentMetadata(
  val userDescription: String? = null
  val lnurlPay: LnurlPayMetadata? = null
)

data class LnurlPayMetadata(
  val url: LNUrl, // the naming may be improved
  val description: String,
  val successAction: LNUrl.PayInvoice.SuccessAction
)
```

This way, we cannot have inconsistent objects where some lnurl fields could be null and we'd have to add a lot of sanity checks downstream when reading data.